pkg/controller/restic_repository_controller.go

```diff
@@ -0,0 +1,227 @@
+// of them so we can check each one for maintenance.
+func (c *resticRepositoryController) enqueueAllRepositories() {
+	c.logger.Debug("resticRepositoryController.enqueueAllRepositories")
+
+	repos, err := c.resticRepositoryLister.List(labels.Everything())
+	if err != nil {
+		c.logger.WithError(errors.WithStack(err)).Error("error listing restic repositories")
+		return
+	}
+
+	for _, repo := range repos {
+		c.enqueue(repo)
+	}
+}
+
+func (c *resticRepositoryController) processQueueItem(key string) error {
+	log := c.logger.WithField("key", key)
+	log.Debug("Running processItem")
+
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		log.WithError(errors.WithStack(err)).Error("error splitting queue key")
+		return nil
+	}
+
+	req, err := c.resticRepositoryLister.ResticRepositories(ns).Get(name)
+	if apierrors.IsNotFound(err) {
+		log.Debug("Unable to find ResticRepository")
+		return nil
+	}
+	if err != nil {
+		return errors.Wrap(err, "error getting ResticRepository")
+	}
+
+	switch req.Status.Phase {
+	case "", v1.ResticRepositoryPhaseNew:
+		// Don't mutate the shared cache
+		return c.initializeRepo(req.DeepCopy(), log)
+	case v1.ResticRepositoryPhaseReady:
+		// run maintenance if due
+		return c.runMaintenanceIfDue(req.DeepCopy(), log)
+	}
+
+	return nil
+}
+
+func (c *resticRepositoryController) initializeRepo(req *v1.ResticRepository, log logrus.FieldLogger) error {
+	log.Debug("resticRepositoryController.initializeRepo")
+
+	// TODO handle the import case (new CR, but repo already exists in object storage)
+
+	// defaulting
+	if req.Spec.MaintenanceFrequency.Duration == 0 {
+		_, err := c.patchResticRepository(req, func(req *v1.ResticRepository) {
+			req.Spec.MaintenanceFrequency = metav1.Duration{Duration: restic.DefaultMaintenanceFrequency}
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	if err := c.resticMgr.InitRepo(req.Name); err != nil {
+		_, err := c.patchResticRepository(req, func(req *v1.ResticRepository) {
+			req.Status.Phase = v1.ResticRepositoryPhaseFailed
+			req.Status.Message = err.Error()
+		})
+		return err
+	}
+
+	_, err := c.patchResticRepository(req, func(req *v1.ResticRepository) {
+		req.Status.Phase = v1.ResticRepositoryPhaseReady
+		req.Status.LastMaintenanceTime = metav1.Time{Time: time.Now()}
+	})
+	return err
+}
+
+func (c *resticRepositoryController) runMaintenanceIfDue(req *v1.ResticRepository, log logrus.FieldLogger) error {
+	log.Debug("resticRepositoryController.runMaintenanceIfDue")
+
+	now := c.clock.Now()
+
+	if !dueForMaintenance(req, now) {
+		log.Debug("not due for maintenance")
+		return nil
+	}
+
+	// TODO if `restic prune` fails, what should we do?
+	// TODO if `restic check` fails (before or after prune), what should we do?
+	log.Debug("pruning repo")
+	if err := c.resticMgr.PruneRepo(req.Name); err != nil {
+		_, err := c.patchResticRepository(req, func(req *v1.ResticRepository) {
+			req.Status.Phase = v1.ResticRepositoryPhaseFailed
```

Comment:
Need to check restic docs. Setting to failed might not be 100% appropriate. Especially if it's a cloud credentials issue (wrong creds - couldn't talk to s3 - shouldn't change state to failed, should it?).

Might need to consider switching to `status.conditions` where 1 condition is `Ready` true/false.